src/main/java/jalf/compiler/Cog.java

```diff
@@ -141,5 +143,30 @@ private Cog hashJoin(Join join, Cog right) {
         return new BaseCog(join, supplier);
     }
 
+    /** Default compilation of `union`. */
+    public Cog union(Union union, Cog right) {
+        AttrList on = union.getUnionAttrList();
+        if (on!=null) {
+            return tupleUnion(union, right);
+        } 
+        else {
+            return null;
+        }
+    }
+
+     /** Compiles a union */
+    private Cog tupleUnion(Union union, Cog right) {
+        // stream compilation: concat + distinct
+        Supplier<Stream<Tuple>> supplier = () ->{
+            Stream<Tuple> leftStream = this.stream();
+            Stream<Tuple> rightStream = right.stream();
+            Optional<Stream<Tuple>> unionStream =Stream.of(leftStream,rightStream)
+               .distinct()
+               .reduce(Stream::concat);
+            return unionStream.orElse(Stream.empty());
+        };
+        return new BaseCog(union, supplier);
+    }
+
```

Comment:
- Merge both methods.
- No null returned by getUnionAttrList() -> enforce at type checking time
- Please check logic for concatening streams (@blambeau)